crypto/bio/bio.c

```diff
@@ -299,13 +308,42 @@ int BIO_write_all(BIO *bio, const void *data, size_t len) {
 }
 
 int BIO_puts(BIO *bio, const char *in) {
-  size_t len = strlen(in);
+  const size_t len = strlen(in);
   if (len > INT_MAX) {
     // |BIO_write| and the return value both assume the string fits in |int|.
     OPENSSL_PUT_ERROR(BIO, ERR_R_OVERFLOW);
     return -1;
   }
-  return BIO_write(bio, in, (int)len);
+  // Check for bwrites here since we use that if bputs is NULL
+  if (bio == NULL || bio->method == NULL || (bio->method->bwrite == NULL &&
+                                            bio->method->bputs == NULL)) {
+    OPENSSL_PUT_ERROR(BIO, BIO_R_UNSUPPORTED_METHOD);
+    return -2;
+  }
+  int ret = 0;
+  if(HAS_CALLBACK(bio)) {
+    ret = (int)bio->callback_ex(bio, BIO_CB_PUTS, in, len, 0, 0L, 1L, NULL);
```

Comment:
For matching behavior we should probably pass `0` here for `len`, as that is the OpenSSL behavior, even though we technically know the length in our emulation.

https://github.com/openssl/openssl/blob/07e4d7f4747005e3ce56423182ad047eb05d8e16/crypto/bio/bio_lib.c#L513